### ISSUE TITLE
Mark final blocks and non-final tx

### ISSUE THREAD

**Issue by ilblackdragon** [near/near-explorer#278]
Currently it's unclear which blocks are final looking at block details or in the home page to latest blocks. We can add "final" tag for them to identify which ones are final.
Alternatively we can show non-final marker for blocks that are non-final yet.

Tx displayed from non-final blocks should also have a tag that they are confirmed but non final.

**Comment by frol** [comment]
Implementation notes:

1. Let's expose `get-finality-stats` via WAMP (backend/src/wamp.js)
  a. Use the new-ish Block API to query block with finality=final (https://github.com/nearprotocol/docs/pull/330). near-api-js does not expose the new API, so it is better to contribute this new API support to near-api-js first.
  b. return the `header.height` as "final_block_height"
2. Query the new WAMP method from the relevant components (transactions list, transaction etails, blocks list, block details) and pass the "isFinal" flag (`block.height <= finalBlockHeight`) down to the dumb component displaying the marker on non-final transactions and blocks.

**Comment by bowenwang1996** [comment]
@frol I don't think it needs to be that complicated. Every block has `last_final_height` in their header and since you already index blocks, you will know which blocks are supposedly final. However, there is a caveat: the blocks whose height are smaller than `last_final_height` of the head isn't necessarily final because they maybe on a different chain. But for practical purposes I think this is sufficient.

**Comment by frol** [comment]
@bowenwang1996 We don't store that information anywhere. To implement this, we will need to come up with some storage for such stats (either global in-memory variable, or key-value SQLite table). Also, the sync process is not sequential, so we should also keep that in mind, so we cannot blindly update the value with the last fetched block.

Thus, querying RPC is just the most reliable way of doing this.

**Comment by bowenwang1996** [comment]
@frol makes sense, but if you are willing to keep the head of the chain then you don't need to do additional queries. Ultimately this info is only good for the data that you have, so if you keep the head of the chain you don't need additional queries (even if you are out of sync at the moment).

**Comment by icerove** [comment]
```
{
  "jsonrpc": "2.0",
  "id": "dontcare",
  "result": {
    "author": "node3",
    "header": {
      "height": 2003340,
      "epoch_id": "3mHzgVWca7GgBc9kQTUY1neKDqzUhkRbpUrMFs1pbWXL",
      "next_epoch_id": "ghpXGxdeLYR1BHvEDJwXF6LYhBsZjytvMkutFyHCtR5",
      "hash": "pUgKX4nT2P6p9aLLMyYh59HQ89i2svbXMsqjdMbdm8F",
      "prev_hash": "FRvF7tvmSAvv4S4vBqQJW75wNfHctU9kGCzAqTp32j78",
      "prev_state_root": "2xZks5S5zih4bW11A6D5YaQi8sfPy3dEkWWXbKSrPe2y",
      "chunk_receipts_root": "9ETNjrt6MkwTgSVMMbpukfxRshSD1avBUUa4R4NuqwHv",
      "chunk_headers_root": "2yKDs8h6UNVDCgLsfCbRck4GibjFAQTA9Y6HS4Aghegp",
      "chunk_tx_root": "7tkzFg8RHBmMw1ncRJZCCZAizgq4rwCftTKYLce8RU8t",
      "outcome_root": "7tkzFg8RHBmMw1ncRJZCCZAizgq4rwCftTKYLce8RU8t",
      "chunks_included": 1,
      "challenges_root": "11111111111111111111111111111111",
      "timestamp": 1588426142437337600,
      "random_value": "CCg1wPWwmvghmLPY1naVavLyFofuRmwvmVp6t3GS9FgQ",
      "score": 2003339,
      "validator_proposals": [],
      "chunk_mask": [
        true
      ],
      "gas_price": "5000",
      "rent_paid": "0",
      "validator_reward": "0",
      "total_supply": "1013828364958054936724694759912258",
      "challenges_result": [],
      "last_quorum_pre_vote": "FRvF7tvmSAvv4S4vBqQJW75wNfHctU9kGCzAqTp32j78",
      "last_quorum_pre_commit": "3RQoGgRMVXyeFxsNBTpZanGqh3renAYyUvH7HH7xGTZP",
      "last_ds_final_block": "FRvF7tvmSAvv4S4vBqQJW75wNfHctU9kGCzAqTp32j78",
      "next_bp_hash": "6jpyHheNEh6f1Q51ZQyDfSsvfBWPFm28d6UBFFAnaaMv",
      "approvals": [
        [
          "node2",
          "FRvF7tvmSAvv4S4vBqQJW75wNfHctU9kGCzAqTp32j78",
          "3tBxnzb7KECu2VybM9bmSin9JUY1Pb8HYc6FagWBzGZR",
          2003340,
          true,
          "ed25519:2wVmQ3XaRsah78aumVo1rwMewEAghiCgNwFMEJFJT4hWyvf2vC6WVQ57SSS7q6P8wyxrf3neke2d2ypQTcWMMYiX"
        ],
        [
          "node1",
          "FRvF7tvmSAvv4S4vBqQJW75wNfHctU9kGCzAqTp32j78",
          "Gi3yTYjAiXzJAmA9EKBeMUnD32Q912nv93PJesKSuGZT",
          2003340,
          true,
          "ed25519:rx11xEhwPRDEnrRcNaUyQX5omkZSsoZGHp2jrYoLhWZQTumP4e2NnoLCvy8ocdMxvJoAqhm2jBzxmguhhSSg9MJ"
        ],
        [
          "node3",
          "FRvF7tvmSAvv4S4vBqQJW75wNfHctU9kGCzAqTp32j78",
          "8CXfTGtNNAi6U5Tnh29qZ7Am6u9tg4QXoWJ1LuiPPShg",
          2003340,
          true,
          "ed25519:36do9Jb1PHXHqVNZgeZU6KW9uWQwmLwqJt8qFwzRes22URkuKZLgQ7Y5qyWjGCNW7jyMeyeMCSTyApo6HH5DAab2"
        ],
        [
          "node0",
          "FRvF7tvmSAvv4S4vBqQJW75wNfHctU9kGCzAqTp32j78",
          "8CXfTGtNNAi6U5Tnh29qZ7Am6u9tg4QXoWJ1LuiPPShg",
          2003340,
          true,
          "ed25519:4Z6Gup9m4h5Zf6pSXt1UK8X7BjhJcb78RWPJRoaTjutvT77F765yC4aREV81JYe1UuQmm293Y6ctSBCFV8XgBCgb"
        ]
      ],
      "signature": "ed25519:vegMXf9vz3hD3bfSW9JbwoHnGf1WpFzqZGWwEWnTbXNYDTKCrRki8XNMjNd1GCxHKSJv8faZHoY3qTThKsbqMka"
    },
    "chunks": [
      {
        "chunk_hash": "FcsjvpEmDbiaxeGNmZJt4KxnZEd9TBA6htBWBp5y1jg2",
        "prev_block_hash": "FRvF7tvmSAvv4S4vBqQJW75wNfHctU9kGCzAqTp32j78",
        "outcome_root": "11111111111111111111111111111111",
        "prev_state_root": "9Ec16Drf8MQwKMoxkHtNz3Qq5wJpWd3bebuMj5PJyujU",
        "encoded_merkle_root": "DMJ4Ava2GSEqCwUuAFX7mCvAZRWRYPVZ6ABKsubPUE3H",
        "encoded_length": 8,
        "height_created": 2003340,
        "height_included": 2003340,
        "shard_id": 0,
        "gas_used": 0,
        "gas_limit": 1000000000000000,
        "rent_paid": "0",
        "validator_reward": "0",
        "balance_burnt": "0",
        "outgoing_receipts_root": "H4Rd6SGeEBTbxkitsCdzfu9xL9HtZ2eHoPCQXUeZ6bW4",
        "tx_root": "11111111111111111111111111111111",
        "validator_proposals": [],
        "signature": "ed25519:3PtiM9fFpkchpvavqQUHpPJDHXbRaPj42qNPoKosHafg539eRmNPGUaGMfA9LyjBAtcyLjMXpqWrmv9DVfjug2yM"
      }
    ]
  }
}

```
I see the "last_ds_final_block" is the same as "prev_hash", is that meaning the final block is actually the previous block? the height is this block.height - 1? @frol

**Comment by frol** [comment]
@icerove `last_ds_final_block` (the block hash finalized by DoomSlug [ds]) is *usually* one block behind the latest, but it is not guaranteed to be the case.

**Comment by kcole16** [comment]
Should we make this a status? e.g. "Finalizing"

For transactions:
<img width="144" alt="Screen Shot 2020-05-11 at 3 17 30 PM" src="https://user-images.githubusercontent.com/5624527/81617882-1af99680-939b-11ea-9ac1-659e96b7543d.png">

For blocks:
- Show as "Finalizing" rather than "Completed"

**Comment by frol** [comment]
@kcole16  good point!

**Comment by corwinharrell** [comment]
Here are the component changes/additions for blocks in the finalizing state. For the moment, it makes sense to use the existing pattern found on the transaction details page for status. The block details header would appear as follows, with an addition of a "status" cell:

<img width="1045" alt="Screen Shot 2020-05-22 at 11 50 14 AM" src="https://user-images.githubusercontent.com/5934188/82687987-17c99a80-9c26-11ea-8b66-934c7e57ad90.png">

For the homepage, @jakestutzman had actually designed a similar state that I was able to surface from the design files, so I just made a small tweak and I think it does the job well of differentiating between block states at a glance:

<img width="720" alt="Screen Shot 2020-05-22 at 12 12 37 PM" src="https://user-images.githubusercontent.com/5934188/82688070-3fb8fe00-9c26-11ea-81a6-6a5b655f179f.png">

**Comment by bowenwang1996** [comment]
@corwinharrell looks great! I think there is a small issue here (not necessarily related to design): because the explorer pulls data from different nodes through a load balancer the information might be inconsistent between different queries (for example which block has been finalized). @frol how is this handled in the backend?

**Comment by frol** [comment]
This will hide the execution status (Succeeded or Failed) for the non-finalized transactions :thinking: 

There are the following states:

* Transaction is successful, though not yet finalized
* Transaction is failed, and this is not yet finalized (no changes were applied to the state, but we record this attempt and charged)
* Transaction is successful and finalized
* Transaction is failed and finalized

@SkidanovAlex @bowenwang1996 @ilblackdragon Correct if I am wrong, but it seems that a transaction can produce a different result when it is executed in different forks, e.g. when another transaction slips before the transaction in question, so there might be not enough balance in one fork while there is enough in another fork.

Thus, we need to display both "success"/"fail" (execution status) and "final"/"non-final" (sync status?).

**Comment by frol** [comment]
> the explorer pulls data from different nodes through a load balancer the information might be inconsistent between different queries

It is currently solved with the load-balancer pinning the requests from the same IP to the same load-balanced instance.

With Network Indexer, we are going to use the information from a single node, so there is not going to be a place for inconsistency, but Explorer will only show the information from a single fork point of view.

**Comment by icerove** [comment]
@corwinharrell Looks great for blocks, do you have the design for transactions to display? Since Frol has comments for transactions, we maybe show two information about success and finality

**Comment by icerove** [comment]
@frol I am a little confused about this: 
> This will hide the execution status (Succeeded or Failed) for the non-finalized transactions 
This design is for block detail page and block block, it is not for transaction page, why it will hide status for transaction page?

**Comment by frol** [comment]
Oh, nvm, I was in the wrong context, what I wrote is applicable to the transactions, not blocks.

The blocks design is solid!

**Comment by bowenwang1996** [comment]
> Thus, we need to display both "success"/"fail" (execution status) and "final"/"non-final" (sync status?).

Can we just show "Finalizing" for transactions that are not finalized and "success/fail" only for transactions that have been finalized?

**Comment by icerove** [comment]
@bowenwang1996 we need design because for now transaction detail page has so many columns and information, we could not just add it.

**Comment by corwinharrell** [comment]
> @corwinharrell Looks great for blocks, do you have the design for transactions to display? Since Frol has comments for transactions, we maybe show two information about success and finality

@icerove sorry yes I was focused on displaying the block status first and am finishing up the different instances of transaction states now. I’m experimenting a bit with the table layout for the details page and am working on a few iterations. I’ll share some of those soon to get feedback. Sorry for the delay :)

**Comment by corwinharrell** [comment]
@icerove Not at all! It’s a fun challenge and I think it’s an opportunity for us to create a better hierarchy then what we have now as well as other explorers like etherscan

**Comment by frol** [comment]
> Can we just show "Finalizing" for transactions that are not finalized and "success/fail" only for transactions that have been finalized?

@bowenwang1996  We can, but I thought that these are two orthogonal metrics. Do you think it is fine to merge them? This will create a single metric of three possible states "Finalizing" | "Success" | "Failure" instead of four (Finalizing-Success, Finalizing-Failure, Final-Success, Final-Failure).

**Comment by bowenwang1996** [comment]
@frol for transactions that are not yet finalized I don't think it makes much sense to display whether they are successful. For most users, they would think that "successful" means that the transaction is fully done. Having transactions that are "successful" suddenly disappear from the history is, in my opinion, more damaging than having users wait a bit.

**Comment by frol** [comment]
Now, I agree with @bowenwang1996. Let's make the status: "Finalizing..." | "Succeeded" | "Failed"

**Comment by corwinharrell** [comment]
> Now, I agree with @bowenwang1996. Let's make the transaction status: "Finalizing..." | "Succeeded" | "Failed"

@frol @icerove @kcole16 Just circling back on this after catching up on the discussion. Are we all in agreement with this decision? If so, is my understanding correct that as a result, we do not need any significant design changes to the UI for transactions other than to reflect a new "finalizing" status?

I'll still be iterating on improvements to the hierarchy of the transaction details view, but I think those layout/design changes can fall outside of this issue if we agree that the UI in its current state can accommodate this small addition.

That being said, I'd like to make a suggestion for the language that we use around transactions vs blocks. While blocks and block states are new concepts to users, transactions and their state are a familiar paradigm to most users used to transacting on the web. While I think it makes sense to indicate that a block is "finalizing" in the context of its permanence being solidified on the blockchain, I don't think it is necessary that this language be mirrored for transactions. From a UX perspective, I actually think it would be more beneficial for us to maintain the more familiar terminology of transactions being "processed", before being labeled as a success or failure.

My proposed terminology for state:

**Blocks**: Initializing... > Finalizing... > Finalized / Complete

**Transactions:** Processing... > Succeeded / Failed

Thoughts? cc @kcole16

**Comment by bowenwang1996** [comment]
@corwinharrell what does "initializing" mean for blocks?

**Comment by corwinharrell** [comment]
> @corwinharrell what does "initializing" mean for blocks?

@bowenwang1996 my intent in using that term was to imply “the creation of” a block, before it is finalized. (Perhaps there’s a better term for this or one already being used that I’m unaware of). Perhaps “initiated” vs “initializing” is better, to imply that a block has been created but has not yet been finalized?

Initiated > Finalizing... > Finalized ?

Just thinking out loud here.

**Comment by frol** [comment]
@corwinharrell Once the block is observable, it is already in a "finalizing" state, meaning that we are only waiting till it is finalized. There is only one state before the block is finalized, and we used to refer to it as "finalizing..."

**Comment by corwinharrell** [comment]
@frol ah got it thanks for the clarification. I assumed that the block was observable before it entered the finalizing state which was incorrect. Disregard my comments about block status then, but let me know what you think about transaction status terminology. cc @bowenwang1996

**Comment by bowenwang1996** [comment]
@corwinharrell yeah frol already clarified the situation with blocks. For transactions I think your proposal makes sense.

**Comment by icerove** [comment]
the PR is merged #297
